tiledb/sm/serialization/serializable_subarray.cc

```diff
@@ -0,0 +1,187 @@
+/**
+ * @file serializable_subarray.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines the class SerializableSubarray.
+ */
+
+// clang-format off
+#ifdef TILEDB_SERIALIZATION
+#include <capnp/compat/json.h>
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include "tiledb/sm/serialization/capnp_utils.h"
+#include "serializable_subarray.h"
+#include "serializable_stats.h"
+#endif
+// clang-format on
+
+#include "tiledb/common/common.h"
+#include "tiledb/sm/enums/layout.h"
+#include "tiledb/sm/subarray/range_subset.h"
+#include "tiledb/sm/subarray/subarray.h"
+
+namespace tiledb {
+namespace sm {
+namespace serialization {
+
+#ifdef TILEDB_SERIALIZATION
+
+extern shared_ptr<Logger> dummy_logger;
+
+std::string SerializableSubarray::to_json() {
```

Comment:
Should we consider a `Serializable` abstract base where `to_capn = 0`, `from_capnp = 0` and `to_json` has a common implementation? I'd be a shame to duplicate to_json in any serializable class we add.

Reply:
to_json and to_capnp take in a different type so I'm not sure how that would work...